### PR TITLE
Allow contact position, normal, and mu to be parameters

### PR DESCRIPTION
**Pull request by rdeits**
(improved version of #15 with the "achievable momentum rate" test fixed)

### CODE REVIEW

File: test/controller.jl [hunk 1]
```diff
@@ -1,3 +1,50 @@
...
+            transform(state, FreeVector3D(root_frame(state.mechanism), 0., 0, 1), default_frame(body))
+        end
+    end
+    controller_contact = QPControl.ContactPoint{4}(position, normal, μ, state, model)
+    controller_contact.maxnormalforce[] = 1e3
+
+    # Constrain the contact force to be non-zero for testing
+    @constraint(model, controller_contact.force_local.v == [0.0, 0.0, 1.0])
+
+    @objective(model, Minimize, 0 * dot(controller_contact.force_local.v, controller_contact.force_local.v)) # Need to have an objective to avoid throwing "unsupported objective" errors https://github.com/tkoolen/SimpleQP.jl/issues/62
```
Comment (tkoolen) [inline test/controller.jl]: Can now be removed.

File: test/controller.jl [hunk 2]
```diff
@@ -1,3 +1,50 @@
...
+
+    for θ in linspace(-π, π)
```
Comment (tkoolen) [inline test/controller.jl]: Could you use `Compat.range` instead?

File: test/controller.jl [hunk 3]
```diff
@@ -1,3 +1,50 @@
...
+        set_configuration!(state, [θ])
+        solve!(model)
+        # Sanity check our constraint
+        @test value.(model, controller_contact.force_local.v) ≈ [0.0, 0.0, 1.0]
+
+        # No matter how we rotate the robot, the contact-aligned frame will still
+        # be aligned to the contact normal, which is fixed in world frame. So the
+        # linear component of the contact wrench in world frame will always be
+        # along [0, 0, 1].
+        @test value.(model, linear(controller_contact.wrench_world)) ≈ [0.0, 0.0, 1.0]
```
Comment (tkoolen) [inline test/controller.jl]: `value.(Ref(model), ...)`

File: test/controller.jl (hunk 1 above)
Comment (rdeits) [reply test/controller.jl]: done!

**Comment by tkoolen** [comment]
Any thoughts on what `evalarg` should be renamed to?

File: test/controller.jl (hunk 3 above)
Comment (rdeits) [reply test/controller.jl]: fixed

File: test/controller.jl (hunk 2 above)
Comment (tkoolen) [reply test/controller.jl]: This one got hidden (sorry, that first one shouldn't have been a single comment).

**Comment by rdeits** [comment]
`{Base|Compat}.coalesce()` ?

**Comment by rdeits** [comment]
Also, the allocation test in the notebook is only failing when run as part of the test set, but not when I just manually `@nbinclude` it or when I run the notebook. Maybe a code coverage issue/

**Comment by tkoolen** [comment]
I haven't seen any allocations due to code coverage since 0.5. Do tests pass locally for you?

**Comment by rdeits** [comment]
No, I get the same issue locally

**Comment by rdeits** [comment]
Oh, i bet it's because we're doing `@benchmark controller(...)` when we should be doing `@benchmark $controller(...)` but we're getting lucky about it being optimized away in some cases

**Comment by rdeits** [comment]
Nope :(

**Comment by rdeits** [comment]
Ok, it *does* pass if I just move the notebook test to the beginning of `runtests.jl` instead of the end....

**Comment by tkoolen** [comment]
Strange. I guess just do that for now, I can take a look later.

**Comment by tkoolen** [comment]
Re: `coalesce`: I think that would be too much of a pun. From the documentation:

> Return the first value in the arguments which is not equal to missing, if any. Otherwise return missing

**Comment by tkoolen** [comment]
Do you think the order-dependent behavior in this PR could be related to https://github.com/JuliaLang/julia/issues/28342?

**Comment by rdeits** [comment]
Maybe. But I only see that issue on master, not on 0.6

**Pull request merged**